pkg/parse/status.go

```diff
@@ -0,0 +1,245 @@
+		sss = HelmSourceSpecStatus{
+			Repo:    source.SourceRepo,
+			Chart:   source.SyncDir.SlashPath(),
+			Version: getChartVersionFromCommit(source.SourceRev, commit),
+		}
+	}
+	return sss
+}
+
+// sourceRev will display the source version,
+// but that could potentially be provided to use as a range of
+// versions from which we pick the latest. We should display the
+// version that was actually pulled down if we can.
+// commit is expected to be of the format `chart:version`,
+// so we parse it to grab the version.
+func getChartVersionFromCommit(sourceRev, commit string) string {
+	split := strings.Split(commit, ":")
+	if len(split) == 2 {
+		return split[1]
+	}
+	return sourceRev
+}
+
+// GitSourceSpecStatus is a SourceSpecStatus for the Git SourceType
+type GitSourceSpecStatus struct {
+	Repo     string
+	Revision string
+	Branch   string
+	Dir      string
+}
+
+// Equals returns true if the specified SourceSpecStatus equals this
+// GitSourceSpecStatus, including type and all field values.
+func (g GitSourceSpecStatus) Equals(other SourceSpecStatus) bool {
+	t, ok := other.(GitSourceSpecStatus)
+	if !ok {
+		return false
+	}
+	return t.Repo == g.Repo &&
+		t.Revision == g.Revision &&
+		t.Branch == g.Branch &&
+		t.Dir == g.Dir
+}
+
+// OCISourceSpecStatus is a SourceSpecStatus for the OCI SourceType
+type OCISourceSpecStatus struct {
+	Image string
+	Dir   string
+}
+
+// Equals returns true if the specified SourceSpecStatus equals this
+// OCISourceSpecStatus, including type and all field values.
+func (o OCISourceSpecStatus) Equals(other SourceSpecStatus) bool {
+	t, ok := other.(OCISourceSpecStatus)
+	if !ok {
+		return false
+	}
+	return t.Image == o.Image &&
+		t.Dir == o.Dir
+}
+
+// HelmSourceSpecStatus is a SourceSpecStatus for the Helm SourceType
+type HelmSourceSpecStatus struct {
+	Repo    string
+	Version string
+	Chart   string
+}
+
+// Equals returns true if the specified SourceSpecStatus equals this
+// HelmSourceSpecStatus, including type and all field values.
+func (h HelmSourceSpecStatus) Equals(other SourceSpecStatus) bool {
+	t, ok := other.(HelmSourceSpecStatus)
+	if !ok {
+		return false
+	}
+	return t.Repo == h.Repo &&
+		t.Version == h.Version &&
+		t.Chart == h.Chart
+}
+
+// SourceStatus represents the status of the source stage of the pipeline.
+type SourceStatus struct {
+	// Spec represents the source specification that this status corresponds to.
+	// The spec is stores in the status so we can distinguish if the status
+	// reflects the latest spec or not.
+	Spec       SourceSpecStatus
+	Commit     string
+	Errs       status.MultiError
+	LastUpdate metav1.Time
+}
+
+// Equals returns true if the specified SourceStatus equals this
+// SourceStatus, excluding the LastUpdate timestamp.
+func (gs SourceStatus) Equals(other SourceStatus) bool {
+	return gs.Commit == other.Commit &&
+		status.DeepEqual(gs.Errs, other.Errs) &&
+		(gs.Spec == nil && other.Spec == nil || gs.Spec.Equals(other.Spec))
```

Comment:
Occurs to me that these could panic if gs.Spec is nil. What's the easiest way to fix this?

Reply:
I'd suggest perhaps just breaking this up into multiple if statements instead of the oneliner conditional. I think it will help readability as well.

As a oneliner you could do this with something like `(gs.Spec == nil && other.Spec == nil) || (gs.Spec != nil && gs.Spec.Equals(other.Spec))`

Reply:
Added an isSpecEqual func to handle the nil case so it can be used inline.